### PR TITLE
Implement grid sampler

### LINKED ISSUE #14457: StochasticTools GridSampler
## Reason

To support applications a GridSampler is required that generates rows distributed in parallel.

## Design

GridSampler will accept a range and number of divisions and generate the sample rows distributed in parallel.

```
vars = '0 10 10
           '1 2 0.1'
```

## Impact

This is a new feature.

### PR DESCRIPTION
**Pull request by aeslaughter**
(closes #14457)

### CODE REVIEW

File: modules/stochastic_tools/include/samplers/CartesianProductSampler.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Sampler.h"
+
+class CartesianProductSampler;
```
Comment (bwspenc) [inline modules/stochastic_tools/include/samplers/CartesianProductSampler.h]: So is CartesianProduct the generally accepted name instead of Grid?  It looks like you're using the 'grid' name in the testing.

Comment (aeslaughter) [reply modules/stochastic_tools/include/samplers/CartesianProductSampler.h]: Yes, "Cartesian Product" is the actual name. I will update the stuff to use that name.

**Review by zachmprince — DISMISSED** [review]

**Comment by aeslaughter** [comment]
Github is not linking to the PR for some reason: https://civet.inl.gov/pr/10123/

File: modules/stochastic_tools/doc/content/source/samplers/CartesianProductSampler.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# CartesianProductSampler
+
+!syntax description /Samplers/CartesianProductSampler
+
+## Overview
+
+The Cartesian product creates a sample for all possible combinations for a set of values,
+$Z = A \times B \times \cdots \times L \times \ M$. The calculation is being performed in parallel
+using a "lazy" scheme to all for the $n_th$ value to be computed exactly as follows
```
Comment (bwspenc) [inline modules/stochastic_tools/doc/content/source/samplers/CartesianProductSampler.md]: I'm not quite sure what it meant by: `using a "lazy" scheme to all for the`.  Maybe clarify that.

Comment (bwspenc) [reply modules/stochastic_tools/doc/content/source/samplers/CartesianProductSampler.md]: `$n_th$` -> `$n$th`

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+registerMooseObject("StochasticToolsApp", CartesianProductSampler);
+
+defineLegacyParams(CartesianProductSampler);
+
+InputParameters
+CartesianProductSampler::validParams()
+{
+  InputParameters params = Sampler::validParams();
+  params.addClassDescription("Provides complete cartesian product for the supplied variables.");
```
Comment (bwspenc) [inline modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: cartesian -> Cartesian

File: modules/stochastic_tools/doc/content/source/samplers/CartesianProductSampler.md (hunk 2 above)
Comment (bwspenc) [reply modules/stochastic_tools/doc/content/source/samplers/CartesianProductSampler.md]: I don't doubt that this is mathematically correct, but if I didn't already know what this class was about, I would have no clue what you were talking about from this description. I would mention that this is also known as a grid sampler, try to describe in words what it does, and provide an example.

File: modules/stochastic_tools/doc/content/source/samplers/CartesianProductSampler.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+       B [ [\frac{n}{|C|\cdots|L||M|}] \textrm{mod} |B|, \\
+       \vdots \\
+       L [ [\frac{n}{|M|}] \textrm{mod} |L|, \\
+       M [ [\frac{n}{1}] \textrm{mod} |M| \\
+       \end{bmatrix}
+
+
+## Example Input File Syntax
+
+!! Describe and include an example of how to use the CartesianProductSampler object.
```
Comment (bwspenc) [inline modules/stochastic_tools/doc/content/source/samplers/CartesianProductSampler.md]: Do this.

File: modules/stochastic_tools/include/samplers/CartesianProductSampler.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+public:
+  static InputParameters validParams();
+
+  CartesianProductSampler(const InputParameters & parameters);
+
+protected:
+  /// Return the sample for the given row and column
+  virtual Real computeSample(dof_id_type row_index, dof_id_type col_index) override;
+
+  // Helper object for computing the CartesianProcduct values; this is a pointer because it cannot
```
Comment (bwspenc) [inline modules/stochastic_tools/include/samplers/CartesianProductSampler.h]: Make this a doxygen comment

File: modules/stochastic_tools/include/utils/BootstrapCalculators.h [hunk 6]
```diff
@@ -40,20 +40,36 @@ std::unique_ptr<const BootstrapCalculator> makeBootstrapCalculator(const MooseEn
...
+   * @param data Vector of data from which statistics are to be computed
+   * @param calc Calculator object defining the statistic to be computed
+   * @param is_distributed Flag indicating if the data is distributed in parallel
+   */
+  virtual std::vector<Real> compute(const std::vector<Real> & data,
+                                    const Calculator & calc,
+                                    const bool is_distributed) const = 0;
 
 protected:
+  // Compute Bootstrap estimates of a statistic
```
Comment (bwspenc) [inline modules/stochastic_tools/include/utils/BootstrapCalculators.h]: Are we still requiring all methods and members to be doxygen commented?  If so, make these doxygen comments.

File: modules/stochastic_tools/include/utils/CartesianProduct.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+/* Compute the Cartesian Product of the supplied vectors.
+ * http://phrogz.net/lazy-cartesian-product
+ * https://github.com/iamtheburd/lazy-cartesian-product-python/blob/master/LazyCartesianProduct.py
+ */
+class CartesianProduct
+{
+public:
+  CartesianProduct(const std::vector<std::vector<Real>> & items);
+
+  // Compute the complete Cartesian product matrix
```
Comment (bwspenc) [inline modules/stochastic_tools/include/utils/CartesianProduct.h]: doxygen

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  params.addRequiredParam<std::vector<Real>>(
+      "items", "A list of items, each item should include a min, step size, and number of steps.");
```
Comment (bwspenc) [inline modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: The name `item` doesn't really mean anything, and the description would leave a user pretty lost as to what this is. Maybe call it something like `intervals`, `ranges_and_steps`, `ranges_and_divisions`? This is something that should be described in a little more depth in the .md file.

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  if (items.size() % 3 != 0)
+    paramError("items",
+               "The number of numeric items must be divisible by 3; min, max, divisions for each "
+               "item are required.");
+
+  std::vector<std::vector<Real>> grid_items;
+  for (std::size_t i = 0; i < items.size(); i += 3)
+  {
+    if (items[i + 2] != std::floor(items[i + 2]))
+      paramError("items", "The third entry for each item must be an integer.");
```
Comment (bwspenc) [inline modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: Explain why it needs to be an integer ('because it defines the number of steps in the interval')

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (items[i + 2] < 0)
+      paramError("items", "The third entry for each item must be positive.");
```
Comment (bwspenc) [inline modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: Explain why it needs to be positive ('because it defines the number of steps in the interval')

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  const std::vector<Real> & items = getParam<std::vector<Real>>("items");
+  if (items.size() % 3 != 0)
+    paramError("items",
+               "The number of numeric items must be divisible by 3; min, max, divisions for each "
+               "item are required.");
+
+  std::vector<std::vector<Real>> grid_items;
+  for (std::size_t i = 0; i < items.size(); i += 3)
+  {
+    if (items[i + 2] != std::floor(items[i + 2]))
```
Comment (bwspenc) [inline modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: Do you also need a check to make sure the first entry is less than the second, or is that permissible?

File: modules/stochastic_tools/src/utils/BootstrapCalculators.C [hunk 12]
```diff
@@ -56,6 +59,25 @@ BootstrapCalculator::BootstrapCalculator(const libMesh::ParallelObject & other,
               "The supplied levels must be less than one");
 }
 
+std::vector<Real>
```
Comment (bwspenc) [inline modules/stochastic_tools/src/utils/BootstrapCalculators.C]: This seems unrelated to the Cartesian product sampling.  Is that right?  Did you mean to include it in this PR?  Since it has its own separate issue, that's OK, but I just wanted to check.

**Review by bwspenc — DISMISSED** [review]

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C (hunk 11 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: I don't think having the first entry less than the second is necessary. A user should be able to define a reverse sequence.

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C (hunk 8 above)
Comment (aeslaughter) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: I think I will leave "items", but will improve the docs.

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C (hunk 11 above)
Comment (aeslaughter) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: This is permissble.

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C (hunk 8 above)
Comment (bwspenc) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: Whaaaat???  I think 'items' is pretty much the least descriptive name you could come up with for a parameter!  I think I'm going to start naming all of my parameters 'items1', 'items2', 'items3', etc.

Comment (bwspenc) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: Or maybe 'stuff1', 'stuff2', 'stuff3'...

File: modules/stochastic_tools/include/utils/BootstrapCalculators.h (hunk 6 above)
Comment (aeslaughter) [reply modules/stochastic_tools/include/utils/BootstrapCalculators.h]: the bootstrap stuff should be gone now, these were from other work that this was on top of.

File: modules/stochastic_tools/src/samplers/CartesianProductSampler.C (hunk 8 above)
Comment (aeslaughter) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: Wow, I struck a nerve. The problem is that I can't come up with anything that isn't also misleading, your suggestions also don't hit the meaning for me as well.

What about creating multiple parameters:
```
starts = `10 20`
steps = `1  2`
counts = `10 20`
```

Comment (bwspenc) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: OK, maybe I was being a little dramatic!

I think something like what you suggest would be an ideal way to make this more clear.  Do you mean 'steps' rather than 'stops', though?

Comment (zachmprince) [reply modules/stochastic_tools/src/samplers/CartesianProductSampler.C]: Maybe `linear_spacing`? Since it's similar to the `linspace` function in python and Matlab. This would also give the opportunity for defining `log_spacing` as a separate option.

**Review by zachmprince — APPROVED** [review]

**Review by zachmprince — APPROVED** [review]

**Review by permcody — APPROVED** [review]
